### ISSUE TITLE
Possible bug in .WOZ & Quarter Tracks

### ISSUE THREAD

**Issue by burniouf** [AppleWin/AppleWin#1110]
if you do:

_**.9				lda D2.CurrentQTrack
				cmp	D2.TargetQTrack
				bne .3

				lsr						CS if X,Y on

				jsr D2.Wait25600usec

				lda IO.D2.Ph0Off,y
				
				bcc .90

				lda IO.D2.Ph0Off,x
				clc

.90				rts**_

the head is supposed to stay between X and Y phases, within Applewin, it moves back to Half track at X

**Comment by tomcw** [comment]
I think it depends on the state of the adjacent phase magnets.

Can you attach a save-state file (.yaml) just before this code is run.

NB. you can do this precisely from within the AppleWin debugger - ie. F11 (Save state) will take effect from the current opcode you are stopped at.

**Comment by burniouf** [comment]
[YAML.zip](https://github.com/AppleWin/AppleWin/files/8840040/YAML.zip)

**Comment by burniouf** [comment]
BEFORE is before lda IO.D2.Ph0Off,y
BETWEEN is after lda IO.D2.Ph0Off,y
AFTER is after lda IO.D2.Ph0Off,x

**Comment by burniouf** [comment]
BEFORE
        Unit0:
          Drive Connected: true
          Phase: 34
          Phase (precise): 34.500000
          Head Window: 0x0
          Last Stepper Cycle: 0x000000009DEE5C03
          Motor On Cycle: 0x000000009D87D40F
          Spinning: 1000000
BETWEEN
       Unit0:
          Drive Connected: true
          Phase: 34
          Phase (precise): 34.000000
          Head Window: 0x0
          Last Stepper Cycle: 0x000000009DEF0967
          Motor On Cycle: 0x000000009D87D40F
          Spinning: 1000000
AFTER
        Unit0:
          Drive Connected: true
          Phase: 34
          Phase (precise): 34.000000
          Head Window: 0x1
          Last Stepper Cycle: 0x000000009DEF096D
          Motor On Cycle: 0x000000009D87D40F
          Spinning: 1000000

**Comment by tomcw** [comment]
It looks like a bug, since rapidly turning of magnet phases 3 & 2 should not affect the position.

From Sather (UTA2e 9-13):
>"...turn two adjacent phases on ... This is referred to as stepping to a quarter track. The two phases must be turned off in rapid sequence if one of them is phase 1, the drive is unmodified, and data is to be written on the quarter track."

>"After head positioning, the four phases should all be turned off, because the drive will behave as if a write protected disk is installed if phase-1 is left on."

```
D44F:AD E4 D5             LDA $D5E4                   
D452:CD E8 D5             CMP $D5E8                   
D455:D0 D9                BNE $D430                   

D457:4A                   LSR                         
D458:20 EA D2             JSR $D2EA                   

; X=64, Y=66
D45B:B9 80 C0             LDA $C080,Y     ; before
D45E:90 04                BCC $D464       ; between
D460:BD 80 C0             LDA $C080,X                 
D463:18                   CLC             ; after
D464:60                   RTS                         
```

yaml | PC | magnet | track
--|--|--|--
before | D45B |$C / 1100 | 34.5
between | D45E | $4 / 0100 | 34.0
after | D463 | $0 / 0000 | 34.0

**Comment by burniouf** [comment]
it happends only when moving head OUT, when moving head inward, no recalibration happends.

btw, i never properly understood this part of Sather 's sentence : "if one of them is phase 1, the drive is unmodified,"
what is particular with phase 1 beside the line is shared with "write protected return" ?

I tried a lot of variation around this usung STA instead of LDA.....no change..

**Comment by burniouf** [comment]
yes, i confirm moving from inner tracks to track 0. Happends when target Qtrack is ODD.
to reproduce, boot A2OSX.TEST 800k media, mount a blank test.woz in S6D1 then type 'llfh1'
llfh1 is a .sh script that LL format a 3/4 track disk (192k) then verify all blocks reading tracks from 0 to 47 then 47 to 0)

**Comment by tomcw** [comment]
(Testing using AppleWin 1.30.10.0, using the release build and also the debug build via Visual Studio to catch any AppleWin asserts.)

When using the 800K [A2OSX.TEST.po](https://github.com/A2osX/A2osX/blob/master/.Floppies/A2OSX.TEST.po), I get the following error message:
![A2OSX TEST](https://user-images.githubusercontent.com/6696896/173243062-cd78567c-4cc7-4a7c-972b-4c6fbf6a2b82.png)

But if I instead use the 32M [A2OSX.BUILD.po](https://github.com/A2osX/A2osX/blob/master/.Floppies/A2OSX.BUILD.po), it completes okay:
![A2OSX BUILD](https://user-images.githubusercontent.com/6696896/173243118-c781610a-444a-4e26-875d-cb2db8ec2904.png)

I don't think this is important though, since I can visually see that when verifying backwards (from track 47 to 0) for both the 800K and 32M images, the output pauses every 2 tracks (for ~1 sec)... which I assume is the problem, and A2osX is having to retry here, and eventually succeeds?

**Comment by burniouf** [comment]
yes, while "verifying", this is not the FORMAT utility that performs seeking, it is ProDOS Disk 2 driver.
So a bad read because of probably failing to read D5 AA 96...etc....triggers a recalibrate to Track 0, then a SEEK to target track, but this time going INward succeeds setting the head at the proper QTrack

Please keep using BUILD media if you now how to use it ....i may have posted a TEST media with debug code somewhere addressing another bug to a specific user ....

**Comment by burniouf** [comment]
Pushed fresh MEDIAs at GH with code modified everywhere like this:

				bcc .10
				clc						Exit wit CC (recalibrate)

				lda IO.D2.Ph0Off,x
				
.10				lda IO.D2.Ph0Off,y

				rts

**Comment by tomcw** [comment]
Hi @burniouf - I assume AppleWin still has the problem, even with this code modification?
(sorry - no time to test the new MEDIAs at the moment)

btw, I have work-in-progress fix for this that I hope to release soon.

**Comment by burniouf** [comment]
LLFH1 on fresh test.woz : KO while seeking outward track 47 -> 0 while VERIFY
same on a already formatted test.woz.

again i said "possible bug", it means i may have also a "possible bug" in my code. :-)
but it s been a long time now i m working on this 3/4 track crazyness, i rewrote the code several time and always faced this behavior in Applewin.

on real hardware, this is a complete defferent story : i have several drives that works perfectly including booting these 3/4 track media.....and on some older model (Disk ][), i noticed a behavior similar to this but only on few tracks : usually something like 5, 9, 13 and 41, and always OUTward direction...INward always works perfectly.
here comes this misterious sentence from Sather s book one of you mentioned earlier : "if one of 2 is ph1 the drive is unmodified" which i never understood well.... :-(
5 9 13 are Qtrack 15 27 39....when MODed 4, always 3 = ph1+ph2 ON....that where i see a link with Sather.....but why OUTward only ????

**Comment by tomcw** [comment]
>"The two phases must be turned off in rapid sequence if one of them is phase 1, the drive is unmodified, and data is to be written on the quarter track."

My understanding of that Sather sentence is:
- "The two phases must be turned off in rapid sequence" - ie. if you are too slow (turn one off, leave one on), then the magnetic attraction will turn the cog and move the head off the quarter track.
- since leaving phase 1 on enable writes protect, then you need phase 1 off if you intend to write to the track (ie. 50% of all quarter tracks will have phase 1 on).
- "the drive is unmodified" - probably modified drives don't enable write protect when phase 1 is on.

Basically he was just being very economical with his words to say what I've said in a much more verbose way.

**Comment by tomcw** [comment]
>again i said "possible bug"

Well, we think there's definitely an issue with AppleWin, as it doesn't account for "The two phases must be turned off in rapid sequence".

For AppleWin the bug is OUTward only (47 -> 0) just because of how the cog/head movement is implemented in the emulator.

**Comment by tomcw** [comment]
With the new A2OSX.BUILD.po, I don't get the pause with 1.30.10.0.

But when stepping the code:

D45D: LDA $C080,X ; C0E4
D460: LDA $C080,Y ; C0E2

PC | magnet | phase precise
--|--|--
D45C | $6 / 0110 | 61.5
D45D | $2 / 0010 | 62.0
D460 | $0 / 0000 | 62.0

So AppleWin 1.30.10.0 still has the bug, as turning the two phases off in rapid sequence should leave it at phase 61.5.

NB. If shouldn't matter which order you turn the phases off, but in AppleWin it makes a difference. I'll get this fixed.

**Comment by burniouf** [comment]
![image](https://user-images.githubusercontent.com/8941204/174425365-f1f40864-c2a1-43de-b6a4-913e1b86e76f.png)

**Comment by burniouf** [comment]
![image](https://user-images.githubusercontent.com/8941204/174425406-84b31e46-2a5d-4f04-b891-fabfb56298d7.png)
you can now use PRODOS.FX / BASIC.FX media for testing.... allows you direct access to any conbination of track count / stepping

**Comment by tomcw** [comment]
@burniouf - here's the 1.30.11.0 [release](https://github.com/AppleWin/AppleWin/releases/tag/v1.30.11.0) with support for two adjacent phases turned off in rapid sequence (and the cog doesn't move).

**Comment by burniouf** [comment]
fine !
quick testing showed bit better...will proceed deeper testing...and....resume coding on this 3/4 track crazyness....
thanks a lot for your reactivity :-)

**Comment by tomcw** [comment]
to do:
- [ ] add a regression test for verifying a pre-formatted 3/4 track .woz image
  - forward & backwards
  - turn phases off in either order: n,n+1 & n+1,n
